packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -2222,34 +2224,116 @@ function flushPassiveEffectsImpl() {
-  // fiber, because the root is not part of its own effect list. This could
-  // change in the future.
-  let effect = root.current.firstEffect;
-  while (effect !== null) {
-    if (__DEV__) {
-      setCurrentDebugFiberInDEV(effect);
-      invokeGuardedCallback(null, commitPassiveHookEffects, null, effect);
-      if (hasCaughtError()) {
-        invariant(effect !== null, 'Should be working on an effect.');
-        const error = clearCaughtError();
-        captureCommitPhaseError(effect, error);
+    // It's important that ALL pending passive effect destroy functions are called
+    // before ANY passive effect create functions are called.
+    // Otherwise effects in sibling components might interfere with each other.
+    // e.g. a destroy function in one component may unintentionally override a ref
+    // value set by a create function in another component.
+    // Layout effects have the same constraint.
+
+    // First pass: Destroy stale passive effects.
+    //
+    // Note: This currently assumes there are no passive effects on the root fiber
+    // because the root is not part of its own effect list.
+    // This could change in the future.
+    let effect = root.current.firstEffect;
+    let effectWithErrorDuringUnmount = null;
+    while (effect !== null) {
+      if (__DEV__) {
+        setCurrentDebugFiberInDEV(effect);
+        invokeGuardedCallback(
+          null,
+          commitPassiveHookUnmountEffects,
+          null,
+          effect,
+        );
+        if (hasCaughtError()) {
+          effectWithErrorDuringUnmount = effect;
+          invariant(effect !== null, 'Should be working on an effect.');
+          const error = clearCaughtError();
+          captureCommitPhaseError(effect, error);
+        }
+        resetCurrentDebugFiberInDEV();
+      } else {
+        try {
+          commitPassiveHookUnmountEffects(effect);
+        } catch (error) {
+          effectWithErrorDuringUnmount = effect;
+          invariant(effect !== null, 'Should be working on an effect.');
+          captureCommitPhaseError(effect, error);
+        }
       }
-      resetCurrentDebugFiberInDEV();
-    } else {
-      try {
-        commitPassiveHookEffects(effect);
-      } catch (error) {
-        invariant(effect !== null, 'Should be working on an effect.');
-        captureCommitPhaseError(effect, error);
+      effect = effect.nextEffect;
+    }
+
+    // Second pass: Create new passive effects.
+    //
+    // Note: This currently assumes there are no passive effects on the root fiber
+    // because the root is not part of its own effect list.
+    // This could change in the future.
+    effect = root.current.firstEffect;
+    while (effect !== null) {
+      // Don't run create effects for a Fiber that errored during destroy.
+      // This check is in place to match previous behavior.
+      // TODO: Rethink whether we want to carry this behavior forward.
+      if (effectWithErrorDuringUnmount !== effect) {
```

Comment:
If multiple fibers errored, this check only works for the last one. So it doesn't seem worth it.

Reply:
Fair enough. I would prefer just not carrying this behavior forward anyway. I'll remove it.